File: src/opencode/audit.ts
/**
 * OpenCode 对账引擎 — 本地 SessionData 聚合 vs OpenCode 官方明细求和
 * ponytail: 仅做月度 token/cost 求和对比，localCost 直接用 usage.cost.total 聚合；差异解释为固定文案
 */
import type { Totals } from "../aggregate.ts";
import type { OpenCodeUsageRecord } from "./types.ts";

export interface OpencodeTotals {
  requests: number;
  input: number;
  output: number;
  cacheRead: number;
  cacheWrite: number;
  reasoning: number;
  totalTokens: number;
  cost: number;
}

export interface AuditDiff {
  requests: number;
  tokens: number;
  cost: number;
}

export interface AuditDiffRate {
  requests: number;
  tokens: number;
  cost: number;
}

export interface AuditResult {
  opencodeTotals: OpencodeTotals;
  diff: AuditDiff;
  diffRate: AuditDiffRate;
  comparison: string;
}

/** 计算 OpenCode 记录的总和（按 ADR-0002 totalTokens = input+cacheRead+output） */
export function computeOpencodeTotals(records: OpenCodeUsageRecord[]): OpencodeTotals {
  let input = 0;
  let output = 0;
  let cacheRead = 0;
  let cacheWrite = 0;
  let reasoning = 0;
  let cost = 0;
  for (const r of records) {
    input += typeof r.inputTokens === "number" && Number.isFinite(r.inputTokens) ? r.inputTokens : 0;
    output += typeof r.outputTokens === "number" && Number.isFinite(r.outputTokens) ? r.outputTokens : 0;
    cacheRead += typeof r.cacheReadTokens === "number" && Number.isFinite(r.cacheReadTokens) ? r.cacheReadTokens : 0;
    const cw5 = typeof r.cacheWrite5mTokens === "number" && Number.isFinite(r.cacheWrite5mTokens) ? r.cacheWrite5mTokens : 0;
    const cw1 = typeof r.cacheWrite1hTokens === "number" && Number.isFinite(r.cacheWrite1hTokens) ? r.cacheWrite1hTokens : 0;
    cacheWrite += cw5 + cw1;
    reasoning += typeof r.reasoningTokens === "number" && Number.isFinite(r.reasoningTokens) ? r.reasoningTokens : 0;
    cost += typeof r.cost === "number" && Number.isFinite(r.cost) ? r.cost : 0;
  }
  const totalTokens = input + cacheRead + output;
  return { requests: records.length, input, output, cacheRead, cacheWrite, reasoning, totalTokens, cost };
}

/**
 * 对账：对比本地 SessionData Totals 与 OpenCode 明细
 * - diff = opencode - local
 * - diffRate = diff / opencode（opencode 为 0 时记 0，避免除零）
 * - comparison 为结构性差异固定说明
 */
export function buildAudit(localTotals: Totals, opencodeRecords: OpenCodeUsageRecord[]): AuditResult {
  const opencodeTotals = computeOpencodeTotals(opencodeRecords);
  const diff: AuditDiff = {
    requests: opencodeTotals.requests - localTotals.requests,
    tokens: opencodeTotals.totalTokens - localTotals.totalTokens,
    cost: opencodeTotals.cost - localTotals.cost,
  };
  const diffRate: AuditDiffRate = {
    requests: opencodeTotals.requests === 0 ? 0 : diff.requests / opencodeTotals.requests,
    tokens: opencodeTotals.totalTokens === 0 ? 0 : diff.tokens / opencodeTotals.totalTokens,
    cost: opencodeTotals.cost === 0 ? 0 : diff.cost / opencodeTotals.cost,
  };
  const comparison =
    "本地统计仅含 pi 会话中计入的 assistant 消息（MessageTimeRange），未计入内部 compaction/分支同步等非会话请求及非 pi 客户端请求；OpenCode 官方账单含全部扣费请求，差额为预期结构性差异。localCost 为本地 usage.cost.total 求和（若未定价则可能为 0），opencode 成本为官方 cost 求和。";
  return { opencodeTotals, diff, diffRate, comparison };
}

File: opencode-analyzer/internal/opencode/audit.go
package opencode

const AuditComparisonExplanation = "本地统计含 Pi 会话中的 assistant/toolResult/compaction/branch_summary 四载体，并应用 billable/cost/failed 门控、request/semantic 去重与 fork 复制历史剔除；未计入非 Pi 客户端请求。OpenCode 官方账单含全部扣费请求，差额为预期结构性差异。localCost 为本地 usage.cost.total 求和（缺少定价时可能为 0），opencode 成本为官方 cost 求和。"

func ComputeOpencodeTotals(records []UsageRecord) OpencodeTotals {
	var input, output, cacheRead, cacheWrite, reasoning, cost float64
	for _, r := range records {
		input += r.InputTokens
		output += r.OutputTokens
		cacheRead += r.CacheReadTokens
		cacheWrite += r.CacheWrite5mTokens + r.CacheWrite1hTokens
		reasoning += r.ReasoningTokens

		rawCost := r.Cost
		if rawCost > 100 {
			cost += rawCost / 1e8
		} else {
			cost += rawCost
		}
	}
	totalTokens := input + cacheRead + output
	return OpencodeTotals{
		Requests:    len(records),
		Input:       input,
		Output:      output,
		CacheRead:   cacheRead,
		CacheWrite:  cacheWrite,
		Reasoning:   reasoning,
		TotalTokens: totalTokens,
		Cost:        cost,
	}
}

func BuildAudit(localTotals LocalTotals, opencodeRecords []UsageRecord, year, month int) AuditResult {
	opTotals := ComputeOpencodeTotals(opencodeRecords)

	diff := AuditDiff{
		Requests: opTotals.Requests - localTotals.Requests,
		Tokens:   opTotals.TotalTokens - localTotals.TotalTokens,
		Cost:     opTotals.Cost - localTotals.Cost,
	}

	var reqRate, tokRate, costRate float64
	if opTotals.Requests > 0 {
		reqRate = float64(diff.Requests) / float64(opTotals.Requests)
	}
	if opTotals.TotalTokens > 0 {
		tokRate = diff.Tokens / opTotals.TotalTokens
	}
	if opTotals.Cost > 0 {
		costRate = diff.Cost / opTotals.Cost
	}

	return AuditResult{
		Year:           year,
		Month:          month,
		LocalTotals:    localTotals,
		OpencodeTotals: opTotals,
		Diff:           diff,
		DiffRate: AuditDiffRate{
			Requests: reqRate,
			Tokens:   tokRate,
			Cost:     costRate,
		},
		Comparison: AuditComparisonExplanation,
	}
}
